datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemRadioGroupViewHolderFactory.kt

```diff
@@ -53,45 +60,28 @@ internal object QuestionnaireItemRadioGroupViewHolderFactory :
         } else {
           prefixTextView.visibility = View.GONE
         }
-        val (questionnaireItem, questionnaireResponseItem) = questionnaireItemViewItem
-        val answer = questionnaireResponseItem.answer.singleOrNull()?.valueCoding
+        val questionnaireItem = questionnaireItemViewItem.questionnaireItem
         questionTextView.text = questionnaireItem.localizedTextSpanned
-        radioGroup.removeAllViews()
-        radioGroup.setOnCheckedChangeListener(null)
-        var index = 0
-        questionnaireItemViewItem.answerOption.forEach {
-          radioGroup.addView(
-            RadioButton(radioGroup.context, null, R.attr.radioButtonStyleQuestionnaire).apply {
-              id = index++ // Use the answer option index as radio button ID
-              text = it.displayString
-              layoutParams =
-                ViewGroup.LayoutParams(
-                  ViewGroup.LayoutParams.MATCH_PARENT,
-                  ViewGroup.LayoutParams.WRAP_CONTENT
-                )
-              isChecked = it.valueCoding.equalsDeep(answer)
-            }
-          )
-        }
-        radioGroup.setOnCheckedChangeListener { _, checkedId ->
-          // if-else block to prevent over-writing of "items" nested within "answer"
-          if (questionnaireResponseItem.answer.size > 0) {
-            questionnaireResponseItem.answer.apply {
-              this[0].value = questionnaireItemViewItem.answerOption[checkedId].value
-            }
-          } else {
-            questionnaireResponseItem.answer.apply {
-              clear()
-              add(
-                QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
-                  value = questionnaireItemViewItem.answerOption[checkedId].value
-                }
-              )
-            }
+        radioGroup.removeViews(1, radioGroup.childCount - 1)
+        flow.referencedIds = IntArray(0)
+        val choiceOrientation =
+          questionnaireItem.choiceOrientation ?: ChoiceOrientationTypes.VERTICAL
+        when (choiceOrientation) {
+          ChoiceOrientationTypes.HORIZONTAL -> {
+            flow.setOrientation(Flow.HORIZONTAL)
+            flow.setWrapMode(Flow.WRAP_CHAIN)
           }
-
-          onAnswerChanged(radioGroup.context)
+          ChoiceOrientationTypes.VERTICAL -> {
+            flow.setOrientation(Flow.VERTICAL)
+            flow.setWrapMode(Flow.WRAP_NONE)
+          }
+        }
+        // By default, radio button starts from index 1
+        questionnaireItemViewItem.answerOption.forEachIndexed { index, answerOption ->
+          populateViewWithAnswerOption(answerOption, choiceOrientation, index)
         }
+        val refIds = IntArray(questionnaireItemViewItem.answerOption.size) { it + 1 }
+        flow.referencedIds = refIds
```

Comment:
same here:

I suggest we use the API `View.generateViewId()` to generate view id rather than doing the book keeping ourselves. this way we don't have to keep doing the +1 business which is probably bug-prone.

In the following code, for each answer option, we are generating a view id -> answer option pair, we pass that to the function, and then at the end we collect the view ids to give them to the flow layout as reference ids.

```suggestion
        questionnaireItemViewItem
          .answerOption
          .map { answerOption -> View.generateViewId() to answerOption }
          .onEach { populateViewWithAnswerOption(it.first, it.second, choiceOrientation) }
          .map { it.first }
          .let { flow.referencedIds = it.toIntArray() }
```

Reply:
okay